File: 백준/Gold/2116. 주사위 쌓기/주사위 쌓기.py
import sys

sys.setrecursionlimit(10**6)

n = int(input())
dice = [list(map(int, input().split())) for _ in range(n)]
opposite = (5, 3, 4, 1, 2, 0)

def solve(depth, total, prev_top):
    if depth == len(dice):
        return total

    next_bottom_index = -1
    for i in range(6):
        if dice[depth][i] == prev_top:
            next_bottom_index = i
            break

    top_index = opposite[next_bottom_index]
    side = 0
    for i in range(6):
        if i != next_bottom_index and i != top_index:
            side = max(side, dice[depth][i])

    return solve(depth + 1, total + side, dice[depth][top_index])

ans = 0

for i in range(6):
    max_side = 0
    for j in range(6):
        if i != j and i != opposite[j]:
            max_side = max(max_side, dice[0][j])

    ans = max(ans, solve(1, max_side, dice[0][i]))

print(ans)

File: 백준/Gold/2116. 주사위 쌓기/주사위 쌓기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int[][] dice;
    static int[] opposite = {5, 3, 4, 1, 2, 0};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        dice = new int[n][6];

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < 6; j++) {
                dice[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int ans = 0;
        for (int i = 0; i < 6; i++) {
            int maxSide = 0;
            for (int j = 0; j < 6; j++) {
                if (i == j || i == opposite[j]) continue;
                maxSide = Math.max(maxSide, dice[0][j]);
            }

            ans = Math.max(ans, solve(1, maxSide, dice[0][i]));
        }

        System.out.println(ans);
    }

    private static int solve(int depth, int sum, int prevTop) {
        if (depth == dice.length) {
            return sum;
        }

        int nextBottomIndex = -1;
        for (int i = 0; i < 6; i++) {
            if (dice[depth][i] == prevTop) {
                nextBottomIndex = i;
                break;
            }
        }

        int topIndex = opposite[nextBottomIndex];
        int maxSide = 0;
        for (int i = 0; i < 6; i++) {
            if (i == nextBottomIndex || i == topIndex) continue;
            maxSide = Math.max(maxSide, dice[depth][i]);
        }

        return solve(depth + 1, sum + maxSide, dice[depth][topIndex]);
    }
}